x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/SparseVectorQueryBuilder.java

```diff
@@ -0,0 +1,361 @@
+        out.writeOptionalString(inferenceId);
+        out.writeOptionalString(text);
+        out.writeOptionalWriteable(tokenPruningConfig);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.field(FIELD_FIELD.getPreferredName(), fieldName);
+        if (queryVectors != null) {
+            builder.startObject(QUERY_VECTOR_FIELD.getPreferredName());
+            for (var token : queryVectors) {
+                token.toXContent(builder, params);
+            }
+            builder.endObject();
+        } else {
+            builder.field(INFERENCE_ID_FIELD.getPreferredName(), inferenceId);
+            builder.field(TEXT_FIELD.getPreferredName(), text);
+        }
+        builder.field(PRUNE_FIELD.getPreferredName(), shouldPruneTokens);
+        if (tokenPruningConfig != null) {
+            builder.field(PRUNING_CONFIG_FIELD.getPreferredName(), tokenPruningConfig);
+        }
+        boostAndQueryNameToXContent(builder);
+        builder.endObject();
+    }
+
+    @Override
+    protected Query doToQuery(SearchExecutionContext context) throws IOException {
+        if (queryVectors == null) {
+            return new MatchNoDocsQuery("Empty query vectors");
+        }
+
+        final MappedFieldType ft = context.getFieldType(fieldName);
+        if (ft == null) {
+            return new MatchNoDocsQuery("The \"" + getName() + "\" query is against a field that does not exist");
+        }
+
+        final String fieldTypeName = ft.typeName();
+        if (fieldTypeName.equals(ALLOWED_FIELD_TYPE) == false) {
+            throw new ElasticsearchParseException(
+                "field [" + fieldName + "] must be type [" + ALLOWED_FIELD_TYPE + "] but is type [" + fieldTypeName + "]"
+            );
+        }
+
+        return (this.tokenPruningConfig == null)
+            ? WeightedTokensUtils.queryBuilderWithAllTokens(queryVectors, ft, context)
+            : WeightedTokensUtils.queryBuilderWithPrunedTokens(fieldName, tokenPruningConfig, queryVectors, ft, context);
+    }
+
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
+
+        // Note: We return a WeightedTokensQueryBuilder here instead of a SparseVectorQueryBuilder, so that this
+        // query will work when search.check_ccs_compatibility is set to true. This will eventually be removed
+        // when it is no longer needed for compatibility.
+        TokenPruningConfig weightedTokensPruningConfig = (tokenPruningConfig != null ? tokenPruningConfig
+            : shouldPruneTokens ? new TokenPruningConfig()
+            : null);
+        if (queryVectors != null) {
+            // TODO return `this` and remove call to weighted tokens builder
+            return new WeightedTokensQueryBuilder(fieldName, queryVectors, weightedTokensPruningConfig, List.of(ALLOWED_FIELD_TYPE));
```

Comment:
Agree that we shouldn't be returning the `WeightedTokensQueryBuilder` - let's follow up on the discussion

Reply:
Resolved internally.